ott/core/potentials.py

```diff
@@ -0,0 +1,164 @@
+import abc
+from typing import Any, Callable, Dict, Sequence, Tuple, Union
+
+import jax
+import jax.numpy as jnp
+import jax.tree_util as jtu
+
+from ott.geometry import pointcloud
+
+__all__ = ["EntropicMap", "DualPotentials"]
+Potential_t = Union[jnp.ndarray, Callable[[jnp.ndarray], float]]
+
+
+@jtu.register_pytree_node_class
+class BaseDualPotentials(abc.ABC):
+  """Base class holding the Kantorovich dual potentials.
+
+  Args:
+    f: The first dual potential.
+    g: The second dual potential.
+
+  Notes:
+    Both potentials can be represented either as an :class:`jax.numpy.ndarray`
+    or as a function taking :class:`jax.numpy.ndarray` and returning
+    :class:`float`.
+  """
+
+  def __init__(self, f: Potential_t, g: Potential_t):
+    self._f = f
+    self._g = g
+
+  @abc.abstractmethod
+  def transport(self, vec: jnp.ndarray, forward: bool = True) -> jnp.ndarray:
+    """Transport points using the dual potentials :attr:`f` and :attr:`g`.
+
+    Args:
+      vec: Points to transport, array of shape ``[n, d]``.
+      forward: Whether to transport the points from source  to the target
+        distribution or vice-versa.
+
+    Returns:
+      The transported points.
+    """
+
+  @property
+  def f(self) -> Potential_t:
+    """The first dual potential."""
+    return self._f
+
+  @property
+  def g(self) -> Potential_t:
+    """The second dual potential."""
+    return self._g
+
+  def tree_flatten(self) -> Tuple[Sequence[Any], Dict[str, Any]]:
+    return [self.f, self.g], {}
+
+  @classmethod
+  def tree_unflatten(
+      cls, aux_data: Dict[str, Any], children: Sequence[Any]
+  ) -> "BaseDualPotentials":
+    return cls(*children, **aux_data)
+
+
+@jtu.register_pytree_node_class
+class EntropicMap(BaseDualPotentials):
+  """Entropic map estimator :cite:`pooladian:21`.
+
+  Args:
+    f: The first dual potential.
+    g: The second dual potential.
+    geom: Geometry associated with the dual potentials.
+  """
+
+  def __init__(
+      self, f: jnp.ndarray, g: jnp.ndarray, geom: pointcloud.PointCloud
+  ):
+    assert geom.is_squared_euclidean, \
+        "Entropic map is only implemented for squared Euclidean cost."
+    super().__init__(f, g)
+    self._geom = geom
+
+  def transport(
+      self,
+      vec: jnp.ndarray,
+      forward: bool = True,
+  ) -> jnp.ndarray:
+    vec = jnp.atleast_2d(vec)
+    f = jnp.zeros(vec.shape[0])  # (k, d)
+    if forward:
+      y, g = self._geom.y, self.g  # (m, d), (m,)
+    else:
+      y, g = self._geom.x, self.f  # (n, d), (n,)
+
+    geom = pointcloud.PointCloud(vec, y, epsilon=self.epsilon)  # (k, {n, m})
```

Comment:
while I think it makes sense to be explicit, and implement an `EntropicMap` class on its own, I think sticking to the grad operator would be more elegant, and would help us keep a single "DualPotentials" class (and maybe no need for `BaseDualPotentials`).

To achieve this, one would need to instantiate potential functions from the results (`f,g`) of and inputs (`geom`) of Sinkhorn. This can be done by instantiating, after convergence, two functions:

`function_f = lambda y : epsilon * lse( (-PointCloud(jnp.at_least2d(y), geom.x).cost + self.f ) / epsilon)`
`function_g = lambda x : epsilon * lse( (-PointCloud(jnp.at_least2d(x), geom.y).cost + self.g ) / epsilon)`
and then instantiate `DualPotentials(function_f, function_g)`.

All of the computations when running `.transport` on that, should then be equivalent to what you wrote below, thanks to the `jax.grad` operator which will compute the derivative of `lse`. WDYT?

Reply:
thinking about this, maybe it might be better to separate the function from the data contained in geom, to facilitate vmapping for instance? sort of a `params` value that would contain the `geom.x` and `geom.y` as well as `f` and `g`? so `function_f` has arguments `vec` and `params` and can be efficiently vmapped on `vec` when needed?

Reply:
> All of the computations when running .transport on that, should then be equivalent to what you wrote below, thanks to the jax.grad operator which will compute the derivative of lse. WDYT?

Agreed, this is now done.

> so function_f has arguments vec and params and can be efficiently vmapped on vec when needed?

Think the above-mentioned `function_{f,g}` is fine with vmapping, no? I just use `f/g/x/y` from the closure.

Reply:
yes, i think this is a fundamental misunderstanding of mine.

i am left with the impression that

```
def f(x, a):
  return ...

f_ = lambda x : f(x, a_0) 

# with closure
_ = jax.vmap(f_)(X)

# without closure
_ = jax.vmap(f, in_axes=[0, None])(X, a_0)
```
that the second approach will use substantially less memory, notably if a_0 is very large. But maybe I am just plain wrong :)

Reply:
I don't know Jax internals that well yet, but based on this thread https://github.com/google/jax/issues/10585#issuecomment-1119699642, I think you might right - the contant array will be baked into the IR so that optimization's can be performed.

Reply:
However, I think the current impl. is still OK, otherwise, I will have to override `f/g/_grad_f/_grad_g`, which I think is not so nice in terms of abstraction anymore.

Reply:
I think this is likely to come back and haunt us at some point. I would assume that the flax modules that are currently used in ICNNs will likely do this efficiently. One possible solution could be to have an interface for a default Callable and another for a `nn` module, in which case we can default to their vmap, which seems to handle this.

then we use `to_sinkhorn_potentials` to instantiate a flax (single) layer NN rather than the direct lse formulation.

https://flax.readthedocs.io/en/latest/api_reference/_autosummary/flax.linen.vmap.html